### DISCUSSION TITLE
Non-Model REST API Plugin Development?

### DISCUSSION THREAD

**Discussion by alextremblay** [nautobot/nautobot#1817] [Q&A] [answered]
Hi All

I've just started diving into making a REST API endpoint in my nautobot plugin. I'm trying to make a rest endpoint which is not tied to a nautobot model, and I'm totally lost.

The nautobot [documentation](https://nautobot.readthedocs.io/en/latest/plugins/development/#adding-rest-api-endpoints) only includes examples for building a rest API endpoint using a ModelViewSet on a custom model. I tried following the django_rest_framework [documentation for creating a non-model `APIView`](https://www.django-rest-framework.org/api-guide/views/), but that failed spectacularly. attempts to GET the API endpoint with curl caused pages and pages of stack traces which start with `AttributeError: type object 'Token' has no attribute 'objects'`.

The solution to that problem was to add `rest_framework.authtoken` to `INSTALLED_APPS` in my nautobot_config.py and running `nautobot-server migrate`. 

Trouble is, the documentation in django_rest_framework for non-model views leverages `rest_framework.authentication.TokenAuthentication`, which appears to be completely different from whatever token authentication system nautobot itself is using. Tokens i provision from my superuser account are rejected by my REST API endpoint with an error: `{"detail": "Invalid token."}`. The DEBUG logs i get from these failed requests are seemingly useless:

```
[pid: 3443019|app: 0|req: 16/17] 172.28.0.93 () {64 vars in 1092 bytes} [Thu May 26 17:52:27 2022] GET /api/plugins/utsc/aruba-blacklist/ => generated 34 bytes in 173 msecs (HTTP/2.0 401) 10 headers in 685 bytes (1 switches on core 0)
17:55:08.909 WARNING django.request :
  Unauthorized: /api/plugins/utsc/aruba-blacklist/
Unauthorized: /api/plugins/utsc/aruba-blacklist/
```

At this point, I'm completely lost. Any tips or advice would be really appreciated!

**Comment by bryanculver** [comment, marked as answer]
So `ModelViewSet` eventually inherits from `APIView` (via `ModelViewSet` > `GenericViewSet` > `GenericAPIView` > `APIView`) and nowhere in the chain does it do anything fancy with the `authentication_classes`.

These are set by default via the `settings.py`:

```
REST_FRAMEWORK = {
    "ALLOWED_VERSIONS": REST_FRAMEWORK_ALLOWED_VERSIONS,
    "DEFAULT_AUTHENTICATION_CLASSES": (
        "rest_framework.authentication.SessionAuthentication",
        "nautobot.core.api.authentication.TokenAuthentication",
    ),
#...
```

You should safely be able to remove the `authentication_classes` ~and `permission_classes`~ from their documented example.

EDIT: See below why `permission_classes` must be set.

**Reply by alextremblay** [reply]
Hmmm.... I must be missing something. I tried that, and now I get this when I attempt to GET the endpoint:

```
Traceback (most recent call last):
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/django/views/generic/base.py", line 70, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/rest_framework/views.py", line 497, in dispatch
    self.initial(request, *args, **kwargs)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/rest_framework/views.py", line 415, in initial
    self.check_permissions(request)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/rest_framework/views.py", line 332, in check_permissions
    if not permission.has_permission(request, self):
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/nautobot/core/api/authentication.py", line 63, in has_permission
    return super().has_permission(request, view)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/rest_framework/permissions.py", line 229, in has_permission
    queryset = self._queryset(view)
  File "/home/atremblay/utsc-tools/.venv/lib/python3.10/site-packages/rest_framework/permissions.py", line 205, in _queryset
    assert hasattr(view, 'get_queryset') \
AssertionError: Cannot apply TokenPermissions on a view that does not set `.queryset` or have a `.get_queryset()` method.
```

**Reply by bryanculver** [reply]
Ah!

So AuthN works as expected but AuthZ (permissions) needs to be changed.

Our `TokenPermissions` inherits from `DjangoObjectPermissions`. It's trying to evaluate if the token has the right permissions to perform the REST verbs (see: https://github.com/nautobot/nautobot/blob/v1.3.4/nautobot/core/api/authentication.py#L34)

You'll need to define and specify your own permission class in this instance, see: https://www.django-rest-framework.org/api-guide/permissions/

So correction to my comment above: You don't need to specify `authentication_classes` but you will need to for `permission_classes`.

**Reply by alextremblay** [reply]
Ahhhhhhhh, ok!

Thank you! I got it working now thanks to you

For anyone else who's wondering, here's my working code so far:

```python
# api/views.py
from rest_framework import permissions
from rest_framework.views import APIView
from rest_framework.response import Response
from rest_framework import status

class ArubaBlacklistView(APIView):
    """
    Get the current aggregated stm blacklist (aka wireless auth blacklist)
    From the aruba controllers, or remove an entry from the blacklist
    """
    permission_classes = [permissions.IsAuthenticated]

    def get(self, request, format=None):
        data = [{"mac-address": "blahblahblah"}] # TODO: get the actual data
        return Response(data)

    def post(self, request, format=None):
        data=request.data
        # TODO: implement this method
        return Response(data, status=status.HTTP_201_CREATED)
        # return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
```python
# api/urls.py
from django.urls import path
from .views import ArubaBlacklistView

urlpatterns = [
    path('aruba-blacklist/', ArubaBlacklistView.as_view(), name='aruba-blacklist'),
]
```

And it works exactly as expected!
```console
$ curl -H 'Authorization: Token <my token>' -H "Accept: application/json;" -s https://<my nautobot server>/api/plugins/utsc/aruba-blacklist/ | jq .
[
  {
    "mac-address": "blahblahblah"
  }
]
```
